h5py/h5t.pyx

```diff
@@ -1647,12 +1652,19 @@ cpdef TypeID py_create(object dtype_in, bint logical=0, bint aligned=0):
         of kind "O" representing a string, it would return an HDF5 variable-
         length string type.
     """
-    cdef dtype dt = dtype(dtype_in)
-    cdef char kind = dt.kind
+    cdef dtype dt
+    cdef char kind
 
-    aligned = getattr(dtype_in, "isalignedstruct", aligned)
+    if hasattr(dtype_in, 'id'):
```

Comment:
`id` is quite generic - if we're checking for `Datatype`, it's probably better to do that explicitly with `isinstance()`.

Reply:
That means that low-level will have a dependency on high-level, which I'd like to avoid.

Reply:
This involves a logical dependency on the high-level object anyway - it has to know that such an object might exist, and how to get the low-level object from it. Using `isinstance()` just makes it explicit that what's going on.

I agree with the idea of keeping the high-level API out of the low-level. But in that case, this function shouldn't accept high-level objects at all - it should accept TypeID, not Datatype.

Reply:
Another option is to find out where `Datatype` is being passed in and use `TypeID` instead.

Reply:
Possibly the easiest way to do that is to let `create_opaque` return a TypeID. The high-level `Datatype` class is only described as being used for named types, and it doesn't really do anything extra on top of TypeID.

Or you could find all the places that the high-level API could pass down a Datatype to the low-level and extract the TypeID instead. That would be a more technically clean way of doing it, but also more work.

Reply:
Yeah, almost everything in HDF5 has an `id` so this is too generic to disambiguate

Reply:
Maybe this clause could be moved below the current `elif`